lib/stores/list_store.dart

```diff
@@ -69,32 +108,123 @@ class ListProvider with ChangeNotifier {
       // FIXME: This is ugly. We should use a redux to not have to do these kind of things.
       //  This is enough for now (it works™) but we should definitly fix it later.
       _tasks.asMap().forEach((i, t) {
         if (task.id == t.id) {
           _tasks[i] = task;
         }
       });
+      _buckets.asMap().forEach((i, b) => b.tasks.asMap().forEach((v, t) {
+        if (task.id == t.id){
+          _buckets[i].tasks[v] = task;
+        }
+      }));
       notifyListeners();
+      return task;
     });
   }
+
+  Future<void> addBucket({BuildContext context, Bucket newBucket, int listId}) {
+    notifyListeners();
+    return VikunjaGlobal.of(context).bucketService.add(listId, newBucket)
+        .then((bucket) {
+          _buckets.add(bucket);
+          notifyListeners();
+        });
+  }
+
+  Future<void> updateBucket({BuildContext context, Bucket bucket}) {
+    return VikunjaGlobal.of(context).bucketService.update(bucket)
+        .then((rBucket) {
+          _buckets[_buckets.indexWhere((b) => rBucket.id == b.id)] = rBucket;
+          _buckets.sort((a, b) => a.position.compareTo(b.position));
+          notifyListeners();
+        });
+  }
+  
+  Future<void> deleteBucket({BuildContext context, int listId, int bucketId}) {
+    return VikunjaGlobal.of(context).bucketService.delete(listId, bucketId)
+        .then((_) {
+          _buckets.removeWhere((bucket) => bucket.id == bucketId);
+          notifyListeners();
+        });
+  }
+
+  Future<void> moveTaskToBucket({BuildContext context, Task task, int newBucketId, int index}) async {
+    final sameBucket = task.bucketId == newBucketId;
+    final newBucketIndex = _buckets.indexWhere((b) => b.id == newBucketId);
+    if (sameBucket && index > _buckets[newBucketIndex].tasks.indexWhere((t) => t.id == task.id)) index--;
+
+    _buckets[_buckets.indexWhere((b) => b.id == task.bucketId)].tasks.remove(task);
+    if (index >= _buckets[newBucketIndex].tasks.length)
+      _buckets[newBucketIndex].tasks.add(task);
+    else
+      _buckets[newBucketIndex].tasks.insert(index, task);
+
+    double kanbanPosition;
+    if (_buckets[newBucketIndex].tasks.length == 1) // only task
+      kanbanPosition = 0.0;
+    else if (index == 0) // first task
+      kanbanPosition = _buckets[newBucketIndex].tasks[1].kanbanPosition / 2.0;
+    else if (index == _buckets[newBucketIndex].tasks.length - 1) // last task
+      kanbanPosition = _buckets[newBucketIndex].tasks[index - 1].kanbanPosition + pow(2.0, 16.0);
+    else // in the middle
+      kanbanPosition = (_buckets[newBucketIndex].tasks[index - 1].kanbanPosition
+          + _buckets[newBucketIndex].tasks[index + 1].kanbanPosition) / 2.0;
+
+    task = await VikunjaGlobal.of(context).taskService.update(task.copyWith(
+      bucketId: newBucketId ?? task.bucketId,
+      kanbanPosition: kanbanPosition,
+    ));
+    _buckets[newBucketIndex].tasks[index] = task;
+
+    // make sure first 2 tasks don't have 0 kanbanPosition
```

Comment:
I only figured this out recently, awesome to see you paid attention to this!

Reply:
Yeah it took me awhile to get the task moving right. I actually ended up looking at the frontend code to see how you did it.